lib/jxl/enc_modular.cc

```diff
@@ -273,46 +275,6 @@ float EstimateWPCost(const Image& img, size_t i) {
   return histo_cost + extra_bits;
 }
 
-float EstimateCost(const Image& img) {
-  // TODO(veluca): consider SIMDfication of this code.
-  size_t extra_bits = 0;
-  float histo_cost = 0;
-  HybridUintConfig config;
-  uint32_t cutoffs[] = {0,  1,  3,  5,   7,   11,  15,  23, 31,
-                        47, 63, 95, 127, 191, 255, 392, 500};
-  constexpr size_t nc = sizeof(cutoffs) / sizeof(*cutoffs) + 1;
-  Histogram histo[nc] = {};
-  for (const Channel& ch : img.channel) {
-    const intptr_t onerow = ch.plane.PixelsPerRow();
-    for (size_t y = 0; y < ch.h; y++) {
-      const pixel_type* JXL_RESTRICT r = ch.Row(y);
-      for (size_t x = 0; x < ch.w; x++) {
-        pixel_type_w left = (x ? r[x - 1] : y ? *(r + x - onerow) : 0);
-        pixel_type_w top = (y ? *(r + x - onerow) : left);
-        pixel_type_w topleft = (x && y ? *(r + x - 1 - onerow) : left);
-        size_t max_diff =
-            std::max({left, top, topleft}) - std::min({left, top, topleft});
-        size_t ctx = 0;
-        for (uint32_t c : cutoffs) {
-          ctx += (c > max_diff) ? 1 : 0;
-        }
-        pixel_type res = r[x] - ClampedGradient(top, left, topleft);
-        uint32_t token;
-        uint32_t nbits;
-        uint32_t bits;
-        config.Encode(PackSigned(res), &token, &nbits, &bits);
-        histo[ctx].Add(token);
-        extra_bits += nbits;
-      }
-    }
-    for (auto& h : histo) {
-      histo_cost += h.ShannonEntropy();
-      h.Clear();
-    }
-  }
-  return histo_cost + extra_bits;
-}
-
```

Comment:
Would it make sense to keep the old scalar implementation as a comment in the simd version? It is probably useful documentation.
Or at least add some high-level comment describing what it does (this was missing from the original code but becomes more important if the code becomes less readable). The function is basically estimating the cost based on the entropy of the residuals of an effort 2 style fixed prediction with ClampedGradient and a fixed context model based on a simple context property (one that doesn't actually exist but presumably is somewhat representative).

Reply:
Sure, will re-add as scalar specialization.